From 56cae4ffacb2a896cb33680b8d622f0e11cacc1a Mon Sep 17 00:00:00 2001
From: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
Date: Thu, 9 May 2024 15:41:18 +0800
Subject: [PATCH 081/236] mtk: hostapd: add support for emlsr

1. Processing the EML capability IE in the association request,
and sending the value of EML field to the kernel.

2. Processing the EML Operating Mode Notification frame,
and sending the EML Operating Mode Notification frame if eml_resp is
true.

Command Usage:
eml_resp -
hostapd_cli -i <interface> eml_resp <enable>

Processing the EML capability IE in the association request,
and sending the value of EML field to the kernel.

The original flow send the mcu command to the firmware using the WCID of
the primary link. Therefore, a  is passed in the hostapd layer
to ensure that the EML OMN is processed and sent to the firmware using
the WCID of the receiving link.

Add a check for the station's EML capabilities.
If the station does not support EMLSR,
do not process the EML OMN frame sent by the station.

Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>

Fix build warning log

warnig: passing argument 2 of 'hostapd_drv_set_eml_omn' discards
'const' qualifier from pointer target type [-Wdiscarded-qualifiers]

Signed-off-by: MeiChia Chiu <MeiChia.Chiu@mediatek.com>
---
 hostapd/config_file.c             |   4 +
 hostapd/ctrl_iface.c              |  36 +++++++++
 hostapd/hostapd.conf              |  10 +++
 hostapd/hostapd_cli.c             |   8 ++
 src/ap/ap_config.h                |   2 +
 src/ap/ap_drv_ops.c               |  16 ++++
 src/ap/ap_drv_ops.h               |   2 +
 src/ap/ieee802_11_eht.c           | 121 ++++++++++++++++++++++++++++--
 src/common/ieee802_11_defs.h      |  27 +++++++
 src/common/mtk_vendor.h           |  15 ++++
 src/drivers/driver.h              |  11 +++
 src/drivers/driver_nl80211.c      |  53 +++++++++++++
 src/drivers/driver_nl80211.h      |   1 +
 src/drivers/driver_nl80211_capa.c |   3 +
 14 files changed, 302 insertions(+), 7 deletions(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index 5efe34bcd..b2dacc25d 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -5251,6 +5251,10 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		}
 	} else if (os_strcmp(buf, "eht_bw320_offset") == 0) {
 		conf->eht_bw320_offset = atoi(pos);
+	} else if (os_strcmp(buf, "eml_disable") == 0) {
+		conf->eml_disable = atoi(pos);
+	} else if (os_strcmp(buf, "eml_resp") == 0) {
+		conf->eml_resp = atoi(pos);
 #ifdef CONFIG_TESTING_OPTIONS
 	} else if (os_strcmp(buf, "eht_oper_puncturing_override") == 0) {
 		if (get_u16(pos, line, &bss->eht_oper_puncturing_override))
diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index ebb3f1cb2..bbe01a0d5 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -5051,6 +5051,40 @@ hostapd_ctrl_iface_disable_beacon(struct hostapd_data *hapd, char *value,
 
 }
 
+static int
+hostapd_ctrl_iface_set_eml_resp(struct hostapd_data *hapd, char *value,
+				char *buf, size_t buflen)
+{
+	struct hostapd_data *link;
+	int cnt = 0;
+	u16 *val;
+
+	if (!hostapd_is_mld_ap(hapd))
+		return -1;
+
+	cnt = hostapd_parse_argument_helper(value, &val);
+	if (cnt == -1)
+		goto fail;
+	if (cnt != 1 || val[0] < 0)
+		goto para_fail;
+
+	for_each_mld_link(link, hapd) {
+		link->iconf->eml_resp = val[0];
+		wpa_printf(MSG_ERROR, "Link:%d, Response EML:%d\n",
+			   link->iconf->band_idx, link->iconf->eml_resp);
+	}
+
+	os_free(val);
+
+	return os_snprintf(buf, buflen, "OK\n");
+
+para_fail:
+	os_free(val);
+	wpa_printf(MSG_ERROR, "Input number or value is incorrect\n");
+fail:
+	return os_snprintf(buf, buflen, "FAIL\n");
+}
+
 static int
 hostapd_ctrl_iface_set_csi(struct hostapd_data *hapd, char *cmd,
 					char *buf, size_t buflen)
@@ -5979,6 +6013,8 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "WMM", 3) == 0) {
 		reply_len = hostapd_ctrl_iface_wmm(hapd, buf + 4,
 						   reply, reply_size);
+	} else if (os_strncmp(buf, "EML_RESP ", 9) == 0) {
+		reply_len = hostapd_ctrl_iface_set_eml_resp(hapd, buf + 9, reply, reply_size);
 	} else {
 		os_memcpy(reply, "UNKNOWN COMMAND\n", 16);
 		reply_len = 16;
diff --git a/hostapd/hostapd.conf b/hostapd/hostapd.conf
index bcd313aaa..5e0dd6ba0 100644
--- a/hostapd/hostapd.conf
+++ b/hostapd/hostapd.conf
@@ -1161,6 +1161,16 @@ wmm_ac_vo_acm=0
 # will be used as the AP MLD MAC address.
 #mld_addr=02:03:04:05:06:07
 
+# EML Capabilities
+# 0 = Enable EML capabilities in Multi-Link Control subfield
+# 1 = Disable EML capabilitites in Multi-Link Control subfield
+#eml_disable=0
+
+# EML Operating Mode Notification frame
+# 0 = AP does not send EML Operating Mode Notification frame to the station
+# 1 = AP sends EML Operating Mode Notification frame to the station
+#eml_resp=1
+
 ##### IEEE 802.1X-2004 related configuration ##################################
 
 # Require IEEE 802.1X authorization
diff --git a/hostapd/hostapd_cli.c b/hostapd/hostapd_cli.c
index 86e5c72ef..642ffba29 100644
--- a/hostapd/hostapd_cli.c
+++ b/hostapd/hostapd_cli.c
@@ -1485,6 +1485,12 @@ static int hostapd_cli_cmd_disable_beacon(struct wpa_ctrl *ctrl, int argc,
 	return hostapd_cli_cmd(ctrl, "NO_BEACON", 1, argc, argv);
 }
 
+static int hostapd_cli_cmd_set_eml_resp(struct wpa_ctrl *ctrl, int argc,
+					char *argv[])
+{
+	return hostapd_cli_cmd(ctrl, "EML_RESP", 1, argc, argv);
+}
+
 #ifdef CONFIG_DPP
 
 static int hostapd_cli_cmd_dpp_qr_code(struct wpa_ctrl *ctrl, int argc,
@@ -1932,6 +1938,8 @@ static const struct hostapd_cli_cmd hostapd_cli_commands[] = {
 		" = show mu onoff value in 0-15 bitmap"},
 	{ "no_beacon", hostapd_cli_cmd_disable_beacon, NULL,
 		"<value> 0: Enable beacon, 1: Disable beacon"},
+	{ "eml_resp", hostapd_cli_cmd_set_eml_resp, NULL,
+		"<value> 0: AP does not send EML Operating Mode Notification frame, 1: AP sends EML OMN frame"},
 #ifdef CONFIG_DPP
 	{ "dpp_qr_code", hostapd_cli_cmd_dpp_qr_code, NULL,
 	  "report a scanned DPP URI from a QR Code" },
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index bd8994866..7ef207ba2 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -1265,6 +1265,8 @@ struct hostapd_config {
 	u8 punct_acs_threshold;
 	u8 eht_default_pe_duration;
 	u8 eht_bw320_offset;
+	u8 eml_disable;
+	u8 eml_resp;
 #endif /* CONFIG_IEEE80211BE */
 
 	/* EHT enable/disable config from CHAN_SWITCH */
diff --git a/src/ap/ap_drv_ops.c b/src/ap/ap_drv_ops.c
index 587e86c75..c1dfc9376 100644
--- a/src/ap/ap_drv_ops.c
+++ b/src/ap/ap_drv_ops.c
@@ -1615,6 +1615,22 @@ int hostapd_drv_beacon_ctrl(struct hostapd_data *hapd, u8 beacon_mode)
 	return hapd->driver->beacon_ctrl(hapd->drv_priv, beacon_mode);
 }
 
+int hostapd_drv_set_eml_omn(struct hostapd_data *hapd, const u8 *mac,
+			    struct eml_omn_element *omn_ie)
+{
+	u8 link_id;
+
+	if (!hapd->driver || !hapd->driver->set_eml_omn)
+		return 0;
+
+	if (!hapd->conf->mld_ap)
+		return 0;
+
+	link_id = hapd->mld_link_id;
+
+	return hapd->driver->set_eml_omn(hapd->drv_priv, link_id, mac, omn_ie);
+}
+
 int hostapd_drv_csi_set(struct hostapd_data *hapd, u8 mode, u8 cfg, u8 v1, u32 v2, u8 *mac)
 {
 	if (!hapd->driver || !hapd->driver->csi_set)
diff --git a/src/ap/ap_drv_ops.h b/src/ap/ap_drv_ops.h
index 297535a4c..453241af1 100644
--- a/src/ap/ap_drv_ops.h
+++ b/src/ap/ap_drv_ops.h
@@ -180,6 +180,8 @@ int hostapd_drv_amnt_dump(struct hostapd_data *hapd, u8 amnt_idx, u8 *amnt_dump_
 int hostapd_drv_background_radar_mode(struct hostapd_data *hapd);
 int hostapd_drv_pp_mode_set(struct hostapd_data *hapd);
 int hostapd_drv_beacon_ctrl(struct hostapd_data *hapd, u8 beacon_mode);
+int hostapd_drv_set_eml_omn(struct hostapd_data *hapd, const u8 *mac,
+			    struct eml_omn_element *omn_ie);
 int hostapd_drv_csi_set(struct hostapd_data *hapd, u8 mode, u8 cfg, u8 v1, u32 v2, u8 *mac);
 int hostapd_drv_csi_dump(struct hostapd_data *hapd, void *dump_buf);
 
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index b3bdc97c5..ccb188f98 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -480,9 +480,11 @@ u8 * hostapd_eid_eht_basic_ml_common(struct hostapd_data *hapd,
 	control = MULTI_LINK_CONTROL_TYPE_BASIC |
 		BASIC_MULTI_LINK_CTRL_PRES_LINK_ID |
 		BASIC_MULTI_LINK_CTRL_PRES_BSS_PARAM_CH_COUNT |
-		BASIC_MULTI_LINK_CTRL_PRES_EML_CAPA |
 		BASIC_MULTI_LINK_CTRL_PRES_MLD_CAPA;
 
+	if (!hapd->iconf->eml_disable)
+		control |= BASIC_MULTI_LINK_CTRL_PRES_EML_CAPA;
+
 	/*
 	 * Set the basic Multi-Link common information. Hard code the common
 	 * info length to 13 based on the length of the present fields:
@@ -492,6 +494,9 @@ u8 * hostapd_eid_eht_basic_ml_common(struct hostapd_data *hapd,
 	 */
 	common_info_len = EHT_ML_COMMON_INFO_LEN;
 
+	if (hapd->iconf->eml_disable)
+		common_info_len -= 2; /* EML Capabilities (2) */
+
 	if (include_mld_id) {
 		/* AP MLD ID */
 		control |= BASIC_MULTI_LINK_CTRL_PRES_AP_MLD_ID;
@@ -511,9 +516,11 @@ u8 * hostapd_eid_eht_basic_ml_common(struct hostapd_data *hapd,
 	/* Currently hard code the BSS Parameters Change Count to 0x1 */
 	wpabuf_put_u8(buf, hapd->eht_mld_bss_param_change);
 
-	wpa_printf(MSG_DEBUG, "MLD: EML Capabilities=0x%x",
-		   hapd->iface->mld_eml_capa);
-	wpabuf_put_le16(buf, hapd->iface->mld_eml_capa);
+	if (!hapd->iconf->eml_disable) {
+		wpa_printf(MSG_DEBUG, "MLD: EML Capabilities=0x%x",
+			   hapd->iface->mld_eml_capa);
+		wpabuf_put_le16(buf, hapd->iface->mld_eml_capa);
+	}
 
 	mld_cap = hapd->iface->mld_mld_capa;
 	max_simul_links = mld_cap & EHT_ML_MLD_CAPA_MAX_NUM_SIM_LINKS_MASK;
@@ -833,12 +840,16 @@ static u8 * hostapd_eid_eht_reconf_ml(struct hostapd_data *hapd, u8 *eid)
 
 
 static size_t hostapd_eid_eht_ml_len(struct mld_info *info,
-				     bool include_mld_id, bool include_bpcc)
+				     bool include_mld_id, bool include_bpcc,
+				     u8 eml_disable)
 {
 	size_t len = 0;
 	size_t eht_ml_len = 2 + EHT_ML_COMMON_INFO_LEN;
 	u8 link_id;
 
+	if (eml_disable)
+		eht_ml_len -= 2; /* EML Capabilities (2) */
+
 	if (include_mld_id)
 		eht_ml_len++;
 
@@ -906,7 +917,8 @@ size_t hostapd_eid_eht_ml_beacon_len(struct hostapd_data *hapd,
 				     struct mld_info *info,
 				     bool include_mld_id)
 {
-	return hostapd_eid_eht_ml_len(info, include_mld_id, false);
+	return hostapd_eid_eht_ml_len(info, include_mld_id, false,
+				      hapd->iconf->eml_disable);
 }
 
 
@@ -1934,7 +1946,7 @@ hostapd_send_link_reconf_resp(struct hostapd_data *hapd,
 		 * once mac80211 is fixed to match the standard (or this comment
 		 * be removed if the standard is modified to match
 		 * implementation). */
-		mle_len = hostapd_eid_eht_ml_len(&mld, false, true);
+		mle_len = hostapd_eid_eht_ml_len(&mld, false, true, hapd->iconf->eml_disable);
 		len += mle_len;
 	}
 
@@ -2738,18 +2750,105 @@ out:
 }
 
 
+static void ieee802_11_send_eml_omn(struct hostapd_data *hapd,
+				    const u8 *addr,
+				    struct eml_omn_element *omn_ie,
+				    size_t len)
+{
+	struct wpabuf *buf;
+
+	buf = wpabuf_alloc(2 + len);
+	if (!buf)
+		return;
+
+	wpabuf_put_u8(buf, WLAN_ACTION_PROTECTED_EHT);
+	wpabuf_put_u8(buf, WLAN_PROT_EHT_EML_OPMODE_NOTIF);
+	wpabuf_put_data(buf, omn_ie, len);
+
+	hostapd_drv_send_action(hapd, hapd->iface->freq, 0, addr,
+				wpabuf_head(buf), wpabuf_len(buf));
+
+	wpabuf_free(buf);
+}
+
+
+static void ieee802_11_rx_eml_omn(struct hostapd_data *hapd,
+				  const u8 *addr, const u8 *frm,
+				  size_t len)
+{
+	struct eml_omn_element *omn_ie;
+
+	if (hapd->iconf->eml_disable) {
+		wpa_printf(MSG_ERROR,
+			   "Ignore EML Operating Mode Notification from "
+			   MACSTR
+			   " since EML Capabilities is disabled",
+			   MAC2STR(addr));
+		return;
+	}
+
+	/* EML Operating Mode Notification IE */
+	omn_ie = os_zalloc(sizeof(struct eml_omn_element));
+	if (omn_ie == NULL)
+		return;
+
+	os_memcpy(omn_ie, frm, len);
+
+	if (omn_ie->control & EHT_EML_OMN_CONTROL_EMLMR_MODE) {
+		wpa_printf(MSG_ERROR,
+			   "EML: Ignore EML Operating Mode Fotification from "
+			   MACSTR
+			   " since doesn't support EMLMR",
+			   MAC2STR(addr));
+		goto out;
+	}
+
+	hostapd_drv_set_eml_omn(hapd, addr, omn_ie);
+
+	omn_ie->control &= ~(EHT_EML_OMN_CONTROL_EMLSR_PARA_UPDATE_COUNT |
+			     EHT_EML_OMN_CONTROL_INDEV_COEX_ACTIVITIES);
+
+	if (hapd->iconf->eml_resp) {
+		ieee802_11_send_eml_omn(hapd, addr, omn_ie, len);
+		wpa_printf(MSG_ERROR, "EML: AP send EML Operating Mode Fotification to "
+				       MACSTR,
+				       MAC2STR(addr));
+	}
+out:
+	os_free(omn_ie);
+	return;
+}
+
+
 void ieee802_11_rx_protected_eht_action(struct hostapd_data *hapd,
 					const struct ieee80211_mgmt *mgmt,
 					size_t len)
 {
 	const u8 *payload;
 	u8 action;
+	struct sta_info *sta;
+	struct mld_info *info;
+	size_t plen;
 
 	if (!hapd->conf->mld_ap)
 		return;
 
+	if (len < IEEE80211_HDRLEN + 2)
+		return;
+
 	payload = ((const u8 *) mgmt) + IEEE80211_HDRLEN + 1;
 	action = *payload++;
+	plen = len - IEEE80211_HDRLEN - 2;
+
+	sta = ap_get_sta(hapd, mgmt->sa);
+	if (!sta) {
+		wpa_printf(MSG_DEBUG, "EHT: Station " MACSTR,
+			   " not found for received Protected EHT Action",
+			   MAC2STR(mgmt->sa));
+		return;
+	}
+
+	info = &sta->mld_info;
 
 	switch (action) {
 	case WLAN_PROT_EHT_LINK_RECONFIG_REQUEST:
@@ -2758,6 +2857,14 @@ void ieee802_11_rx_protected_eht_action(struct hostapd_data *hapd,
 			wpa_printf(MSG_INFO,
 				   "MLD: Link Reconf Request processing failed");
 		return;
+	case WLAN_PROT_EHT_EML_OPMODE_NOTIF:
+		if (!info->common_info.eml_capa & EHT_ML_EML_CAPA_EMLSR_SUPP) {
+			wpa_printf(MSG_ERROR, "EHT: Fail, Station does not support EMLSR!");
+			return;
+		}
+
+		ieee802_11_rx_eml_omn(hapd, mgmt->sa, payload, plen);
+		return;
 	}
 
 	wpa_printf(MSG_DEBUG,
diff --git a/src/common/ieee802_11_defs.h b/src/common/ieee802_11_defs.h
index af6036e2d..db26cfb34 100644
--- a/src/common/ieee802_11_defs.h
+++ b/src/common/ieee802_11_defs.h
@@ -796,6 +796,33 @@
 #define WLAN_PROT_EHT_LINK_RECONFIG_REQUEST 11
 #define WLAN_PROT_EHT_LINK_RECONFIG_RESPONSE 12
 
+/* EML Operating Mode Notification frame */
+#define EHT_EML_OMN_CONTROL_EMLSR_MODE 0x1
+#define EHT_EML_OMN_CONTROL_EMLMR_MODE 0x2
+#define EHT_EML_OMN_CONTROL_EMLSR_PARA_UPDATE_COUNT 0x4
+#define EHT_EML_OMN_CONTROL_INDEV_COEX_ACTIVITIES 0x8
+
+/* EMLSR Parameter Update field */
+#define EHT_EML_OMN_EMLSR_PADDING_DELAY_MASK 0x07
+#define EHT_EML_OMN_EMLSR_TRANSITION_DELAY_MASK 0x38
+
+struct eml_omn_element {
+	u8 dialog_token;
+	u8 control;
+	le16 bitmap;
+	union {
+		struct {
+			u8 emlsr_para_update;
+		} STRUCT_PACKED emlsr_info;
+		struct {
+			u8 mcs_map_count_control;
+			u8 mcs_map_bw80[3];
+			u8 mcs_map_bw160[3];
+			u8 mcs_map_bw320[3];
+		} STRUCT_PACKED emlmr_info;
+	} u;
+} STRUCT_PACKED;
+
 /* Radio Measurement capabilities (from RM Enabled Capabilities element)
  * IEEE Std 802.11-2020, 9.4.2.44, Table 9-179 */
 /* byte 1 (out of 5) */
diff --git a/src/common/mtk_vendor.h b/src/common/mtk_vendor.h
index 05a1c6f50..0aaf20883 100644
--- a/src/common/mtk_vendor.h
+++ b/src/common/mtk_vendor.h
@@ -19,6 +19,7 @@ enum mtk_nl80211_vendor_subcmds {
 	MTK_NL80211_VENDOR_SUBCMD_BACKGROUND_RADAR_CTRL = 0xcb,
 	MTK_NL80211_VENDOR_SUBCMD_PP_CTRL = 0xcc,
 	MTK_NL80211_VENDOR_SUBCMD_BEACON_CTRL = 0xcd,
+	MTK_NL80211_VENDOR_SUBCMD_EML_CTRL = 0xd3,
 };
 
 enum mtk_vendor_attr_edcca_ctrl {
@@ -286,6 +287,20 @@ enum mtk_vendor_attr_beacon_ctrl {
 		NUM_MTK_VENDOR_ATTRS_BEACON_CTRL - 1
 };
 
+enum mtk_vendor_attr_eml_ctrl {
+
+	MTK_VENDOR_ATTR_EML_CTRL_UNSPEC,
+
+	MTK_VENDOR_ATTR_EML_LINK_ID,
+	MTK_VENDOR_ATTR_EML_STA_ADDR,
+	MTK_VENDOR_ATTR_EML_CTRL_STRUCT,
+
+	/* keep last */
+	NUM_MTK_VENDOR_ATTRS_EML_CTRL,
+	MTK_VENDOR_ATTR_EML_CTRL_MAX =
+		NUM_MTK_VENDOR_ATTRS_EML_CTRL -1
+};
+
 #define CSI_BW20_DATA_COUNT	64
 #define CSI_BW40_DATA_COUNT	128
 #define CSI_BW80_DATA_COUNT	256
diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index fd8a6a145..248e70c98 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -5556,6 +5556,17 @@ struct wpa_driver_ops {
 	 */
 	int (*beacon_ctrl)(void *priv, u8 beacon_mode);
 
+	/**
+	 * set eml omn - Send the EML Operating Mode
+	 * 		 Notification content to driver
+	 * @priv: Private driver interface data
+	 * @link_id: MLD link id
+	 * @addr: MLD STA address
+	 * @omn_ie: EML OMN content sent by the MLD STA
+	 */
+	int (*set_eml_omn)(void *priv, u8 link_id, const u8 *addr,
+                           struct eml_omn_element *omn_ie);
+
 	/**
 	 * three_wire_ctrl - set three_wire_ctrl mode
 	 * @priv: Private driver interface data
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 0816216a1..7d63950fa 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -2975,6 +2975,9 @@ static int nl80211_action_subscribe_ap(struct i802_bss *bss)
 	/* Vendor-specific Protected */
 	if (nl80211_register_action_frame(bss, (u8 *) "\x7e", 1) < 0)
 		ret = -1;
+	/* Protected EHT */
+	if (nl80211_register_action_frame(bss, (u8 *) "\x25", 1) < 0)
+		ret = -1;
 	/* Vendor-specific */
 	if (nl80211_register_action_frame(bss, (u8 *) "\x7f", 1) < 0)
 		ret = -1;
@@ -6166,6 +6169,14 @@ static int wpa_driver_nl80211_sta_add(void *priv,
 			goto fail;
 	}
 
+	if (params->eml_cap) {
+		wpa_printf(MSG_DEBUG, "  * eml_capa=%u",
+			   params->eml_cap);
+		if (nla_put_u16(msg, NL80211_ATTR_EML_CAPABILITY,
+				params->eml_cap))
+			goto fail;
+	}
+
 	ret = send_and_recv_cmd(drv, msg);
 	msg = NULL;
 	if (ret)
@@ -16165,6 +16176,47 @@ static int nl80211_get_mld_addr(void *priv, u8 *addr)
 
 	return 0;
 }
+
+static int nl80211_set_eml_omn(void *priv, u8 link_id, const u8 *addr,
+			       struct eml_omn_element *omn_ie)
+{
+	struct i802_bss *bss = priv;
+	struct wpa_driver_nl80211_data *drv = bss->drv;
+	struct nl_msg *msg;
+	struct nlattr *data;
+	int ret = -ENOBUFS;
+
+	if (!drv->mtk_eml_vendor_cmd_avail) {
+		wpa_printf(MSG_ERROR,
+			   "nl80211: Driver does not support setting EML control");
+		return 0;
+	}
+
+	if (!(msg = nl80211_drv_msg(drv, 0, NL80211_CMD_VENDOR)) ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_ID, OUI_MTK) ||
+	    nla_put_u32(msg, NL80211_ATTR_VENDOR_SUBCMD,
+			MTK_NL80211_VENDOR_SUBCMD_EML_CTRL) ||
+	    !(data = nla_nest_start(msg, NL80211_ATTR_VENDOR_DATA)) ||
+	    nla_put_u8(msg, MTK_VENDOR_ATTR_EML_LINK_ID, link_id) ||
+	    nla_put(msg, MTK_VENDOR_ATTR_EML_STA_ADDR, ETH_ALEN, addr) ||
+	    nla_put(msg, MTK_VENDOR_ATTR_EML_CTRL_STRUCT,
+		    sizeof(struct eml_omn_element), omn_ie))
+		goto fail;
+
+
+	nla_nest_end(msg, data);
+	ret = send_and_recv_cmd(drv, msg);
+
+	if (ret)
+		wpa_printf(MSG_ERROR, "Failed to set EML OMN ctrl. ret = %d (%s)",
+			   ret, strerror(-ret));
+
+	return ret;
+
+fail:
+	nlmsg_free(msg);
+	return ret;
+}
 #endif
 
 static int
@@ -16533,6 +16585,7 @@ const struct wpa_driver_ops wpa_driver_nl80211_ops = {
 	.mu_ctrl = nl80211_mu_ctrl,
 	.mu_dump = nl80211_mu_dump,
 	.beacon_ctrl = nl80211_beacon_ctrl,
+	.set_eml_omn = nl80211_set_eml_omn,
 #ifdef CONFIG_DPP
 	.dpp_listen = nl80211_dpp_listen,
 #endif /* CONFIG_DPP */
diff --git a/src/drivers/driver_nl80211.h b/src/drivers/driver_nl80211.h
index 54464926e..2a22e2fae 100644
--- a/src/drivers/driver_nl80211.h
+++ b/src/drivers/driver_nl80211.h
@@ -213,6 +213,7 @@ struct wpa_driver_nl80211_data {
 	unsigned int mtk_pp_vendor_cmd_avail:1;
 	unsigned int mtk_beacon_ctrl_vendor_cmd_avail:1;
 	unsigned int mtk_csi_vendor_cmd_avail:1;
+	unsigned int mtk_eml_vendor_cmd_avail:1;
 
 	u8 extra_bss_membership_selectors[8];
 
diff --git a/src/drivers/driver_nl80211_capa.c b/src/drivers/driver_nl80211_capa.c
index 4f181e2d3..f1d95c7d6 100644
--- a/src/drivers/driver_nl80211_capa.c
+++ b/src/drivers/driver_nl80211_capa.c
@@ -1184,6 +1184,9 @@ static int wiphy_info_handler(struct nl_msg *msg, void *arg)
 				case MTK_NL80211_VENDOR_SUBCMD_CSI_CTRL:
 					drv->mtk_csi_vendor_cmd_avail = 1;
 					break;
+				case MTK_NL80211_VENDOR_SUBCMD_EML_CTRL:
+					drv->mtk_eml_vendor_cmd_avail = 1;
+					break;
 				}
 			}
 
-- 
2.45.2

